File: analysis/0_preprocessing.py
import json


import numpy as np
import pandas as pd


# Get files from OSF ======================================================
def osf_listfiles(data_subproject="", token="", after_date=None):
    """Function to connect and access files from an OSF project."""
    try:
        import osfclient
    except ImportError:
        raise ImportError("Please install 'osfclient' (`pip install osfclient`)")
    osf = osfclient.OSF(token=token).project(data_subproject)  # Connect to project
    storage = [s for s in osf.storages][0]  # Access storage component
    files = [
        {
            "name": file.name.replace(".csv", ""),
            "date": pd.to_datetime(file.date_created),
            "url": file._download_url,
            "size": file.size,
            "file": file,
        }
        for file in storage.files
    ]

    if after_date is not None:
        date = pd.to_datetime(after_date, format="%d/%m/%Y", utc=True)
        files = [f for f, d in zip(files, [f["date"] > date for f in files]) if d]
    return files


# Connect to OSF and get files --------------------------------------------
token = ""  # Paste OSF token here to access private repositories
files = osf_listfiles(
    token=token,
    data_subproject="maf92",  # Data subproject ID
    after_date="19/01/2024",
)

# Loop through files ======================================================
# Initialize empty dataframes
alldata = pd.DataFrame()
alldata_mist = pd.DataFrame()

for i, file in enumerate(files):
    print(f"File N°{i+1}/{len(files)}")

    if (
        "Participant" in alldata.columns
        and file["name"] in alldata["Participant"].values
    ):
        continue

    download_ok = False
    while download_ok == False:
        data = pd.read_csv(file["file"]._get(file["url"], stream=True).raw)
        if len(data) > 0:
            download_ok = True

    # Participant ----------------------------------------------------------
    data["screen"].unique()

    # Browser info -------------------------------------------------------
    browser = data[data["screen"] == "browser_info"].iloc[0]

    if isinstance(browser["prolific_id"], str) is False:
        continue

    df = pd.DataFrame(
        {
            "Participant": file["name"],
            "Prolific_ID": browser["prolific_id"],
            "Experiment_Duration": data["time_elapsed"].max() / 1000 / 60,
            "Date_OSF": file["date"],
            "Date": browser["date"],
            "Time": browser["time"],
            "Browser": browser["browser"],
            "Mobile": browser["mobile"],
            "Platform": browser["os"],
            "Screen_Width": browser["screen_width"],
            "Screen_Height": browser["screen_height"],
        },
        index=[0],
    )

    # Demographics -------------------------------------------------------
    demo1 = data[data["screen"] == "demographics_1"].iloc[0]
    demo1 = json.loads(demo1["response"])

    df["Gender"] = demo1["gender"]

    # Education
    edu = demo1["education"]
    edu = "Bachelor" if "bachelor" in edu else edu
    edu = "Master" if "master" in edu else edu
    edu = "Doctorate" if "doctorate" in edu else edu
    edu = "High School" if "High school" in edu else edu
    df["Education"] = edu

    demo2 = data[data["screen"] == "demographics_2"].iloc[0]
    demo2 = json.loads(demo2["response"])
    df["Age"] = demo2["age"]
    df["Ethnicity"] = demo2["ethnicity"]

    # Intervention ----------------------------------------------------------------
    intervention = data[data["screen"] == "intervention"].iloc[0]
    df["Intervention_Duration"] = intervention["rt"] / 1000 / 60

    # BFI ----------------------------------------------------------------
    bfi = data[data["screen"] == "questionnaire_BFI10"].iloc[0]

    df["BFI_Duration"] = bfi["rt"] / 1000 / 60

    bfi = json.loads(bfi["response"])
    for item in bfi:
        df[item] = float(bfi[item])

    # VSA-----------------------------------------------------
    vsa = data[data["screen"] == "questionnaire_VSA"].iloc[0]
    df["VSA_Duration"] = vsa["rt"] / 1000 / 60

    vsa = json.loads(vsa["response"])
    for item in vsa:
        df[item] = float(vsa[item])

    # ANES-----------------------------------------------------
    anes = data[data["screen"] == "questionnaire_ANES"].iloc[0]
    df["ANES_Duration"] = anes["rt"] / 1000 / 60

    anes = json.loads(anes["response"])
    for item in anes:
        # Split the string at the first '.' and take the first part (the number), then convert it to float
        numeric_part = anes[item].split(".")[0].strip()  # '2. Slightly Liberal' -> '2'
        df[item] = float(numeric_part)

    # GCBS15-----------------------------------------------------
    gcbs = data[data["screen"] == "questionnaire_GCBS15"].iloc[0]
    df["GCBS_Duration"] = gcbs["rt"] / 1000 / 60

    gcbs = json.loads(gcbs["response"])
    for item in gcbs:
        df[item] = float(gcbs[item])

    # intervention-----------------------------------------------------
    badnews = data[data["screen"] == "intervention"].iloc[0]
    df["BadNews_Duration"] = badnews["rt"] / 1000 / 60

    # veracity confidence-----------------------------------------------------
    confidence = data[data["screen"] == "veracity_confidence"].iloc[0]
    df["Confidence_Duration"] = confidence["rt"] / 1000 / 60

    # badnews questions-----------------------------------------------------
    if not data[data["screen"] == "questionnaire_badnews"].empty:
        badnews_questions = data[data["screen"] == "questionnaire_badnews"].iloc[0]
        df["Badnews_Questions_Duration"] = badnews_questions["rt"] / 1000 / 60
        df["Badnews_Questions_Followers"] = badnews_questions["followers"]
        df["Badnews_Questions_Favorite_Part"] = badnews_questions["favorite_part"]
        df["Badnews_Questions_Play_Again"] = badnews_questions["play_again"]
        df["Badnews_Questions_recommend"] = badnews_questions["recommend"]
    else:
        df["Badnews_Questions_Duration"] = None  # or another placeholder value
        df["Badnews_Questions_Followers"] = None  # or another placeholder value
        df["Badnews_Questions_Favorite_Part"] = None  # or another placeholder value
        df["Badnews_Questions_Play_Again"] = None  # or another placeholder value
        df["Badnews_Questions_recommend"] = None  # or another placeholder value

    # tetris-----------------------------------------------------
    if not data[data["screen"] == "questionnaire_tetris"].empty:
        tetris_questions = data[data["screen"] == "questionnaire_tetris"].iloc[0]
        df["Tetris_Questions_Duration"] = tetris_questions["rt"] / 1000 / 60
        df["Tetris_Questions_Score"] = tetris_questions["high_score"]
        df["Tetris_Questions_Favorite_Part"] = tetris_questions["favorite_part"]
        df["Tetris_Questions_Play_Again"] = tetris_questions["play_again"]
        df["Tetris_Questions_Recommend"] = tetris_questions["recommend"]
    else:
        df["Tetris_Questions_Duration"] = None  # or another placeholder value
        df["Tetris_Questions_Score"] = None  # or another placeholder value
        df["Tetris_Questions_Favorite_Part"] = None  # or another placeholder value
        df["Tetris_Questions_Play_Again"] = None  # or another placeholder value
        df["Tetris_Questions_Recommend"] = None  # or another placeholder value

    # Post-intervention confidence questionnaire
    veracity_confidence = data[data["screen"] == "veracity_confidence"].iloc[0]
    df["Veracity_Confidence"] = veracity_confidence["veracity_confidence"]

    
    # consent-----------------------------------------------------
    consent = data[data["screen"] == "consent"].iloc[0]
    df["Consent_Duration"] = consent["rt"] / 1000 / 60

    # waitdatasaving-----------------------------------------------------
    wds = data[data["screen"] == "waitdatasaving"].iloc[0]
    df["Waitdatasaving_Duration"] = wds["rt"] / 1000 / 60

    # MOCRI BOLD------------------------------------------------
    mocri_bold = data[data["screen"] == "questionnaire_MOCRI_BOLD_12"].iloc[0]
    df["MOCRI_BOLD_Duration"] = mocri_bold["rt"] / 1000 / 60

    mocri_bold = json.loads(mocri_bold["response"])
    for item in mocri_bold:
        df[f"MOCRI_BOLD_{item}"] = float(mocri_bold[item])

    # MOCRI NONBOLD
    mocri_nonbold = data[data["screen"] == "questionnaire_MOCRI_NONBOLD_12"].iloc[0]
    df["MOCRI_NONBOLD_Duration"] = mocri_nonbold["rt"] / 1000 / 60

    mocri_nonbold = json.loads(mocri_nonbold["response"])
    for item in mocri_nonbold:
        df[f"MOCRI_NONBOLD_{item}"] = float(mocri_nonbold[item])

    # MIST ----------------------------------------------------------------

    # Pre
    df_mist_pre = pd.DataFrame({"Participant": df["Participant"].values[0]}, index=[0])
    mist_pre = data[data["screen"] == "questionnaire_mist_pre"].iloc[0]

    df["MIST_Duration_Pre"] = mist_pre["rt"] / 1000 / 60

    mist_pre = json.loads(mist_pre["response"])
    for item in mist_pre:
        df_mist_pre[item] = float(mist_pre[item])
    df_mist_pre = df_mist_pre.T
    df_mist_pre = df_mist_pre.drop("Participant", axis=0)
    df_mist_pre.columns = ["MIST"]  # Rename column
    df_mist_pre["Condition"] = "Pretest"
    df_mist_pre = df_mist_pre.reset_index(names="Item")

    # Post
    df_mist_post = pd.DataFrame({"Participant": df["Participant"].values[0]}, index=[0])
    mist_post = data[data["screen"] == "questionnaire_mist_post"].iloc[0]

    df["MIST_Duration_Post"] = mist_post["rt"] / 1000 / 60
    mist_post = json.loads(mist_post["response"])
    for item in mist_post:
        df_mist_post[item] = float(mist_post[item])
    df_mist_post = df_mist_post.T
    df_mist_post = df_mist_post.drop("Participant", axis=0)
    df_mist_post.columns = ["MIST"]  # Rename column
    df_mist_post["Condition"] = "Posttest"
    df_mist_post = df_mist_post.reset_index(names="Item")

    # Concatenate
    df_mist = pd.concat([df_mist_pre, df_mist_post], axis=0, ignore_index=True)
    df_mist["Participant"] = df["Participant"].values[0]

    # Add to alldata
    alldata = pd.concat([alldata, df], axis=0, ignore_index=True)
    alldata_mist = pd.concat(
        [alldata_mist, df_mist[["Participant", "Item", "Condition", "MIST"]]],
        axis=0,
        ignore_index=True,
    )
print("Done!")



# Quality control =========================================================
def update_log(log, prolific_id, reject=False, alldata=alldata):
    ppt = alldata.loc[
        (~alldata.Prolific_ID.isna()) & (alldata.Prolific_ID == prolific_id),
    ]["Participant"].values[0]
    if not isinstance(ppt, str):
        raise ValueError(f"Participant {prolific_id} not found.")

    ppt = pd.DataFrame(
        {"Participant": [ppt], "Paid": [not reject], "Date": pd.Timestamp.now()}
    )
    log = pd.concat([log, ppt], axis=0, ignore_index=True)
    # Remove duplicates
    log = log.drop_duplicates(subset="Participant", keep="first")
    log.to_csv("../data/payment_log.csv", index=False)
    if reject:
        print(f"X Participant {prolific_id} rejected.")
    else:
        print(f"V Participant {prolific_id} was paid.")
    return log


# Load log and data
log = pd.read_csv("../data/payment_log.csv")
new = alldata.loc[~alldata.Prolific_ID.isna(),]
new = new.loc[
    ~new.Participant.isin(log.Participant), ["Prolific_ID", "Intervention_Duration"]
]
new
new.head(40)

# Update log manually
log = update_log(log, "6620d6cae03db8263ba0f4c0", reject=True)

alldata[["Prolific_ID",  "Intervention_Duration"]]
# Save data ==============================================================
alldata = alldata.drop("Prolific_ID", axis=1)
alldata.to_csv("../data/rawdata.csv", index=False)
alldata_mist.to_csv("../data/rawdata_mist.csv", index=False)

File: analysis/0_preprocessing.R
library(jsonlite)
library(progress)
library(stringr)
library(dplyr)


# Define local path to participant files (modify as needed)
path <- "C:/Users/domma/Box/Data/FakeNewsIntervention/"
# path <- "C:/Users/dmm56/Box/Data/FakeNewsIntervention/"



# List all CSV files in the local directory
files <- c(
  list.files(paste0(path, "Prolific/"), full.names = TRUE, pattern = "*.csv"),
  list.files(paste0(path, "UK/"), full.names = TRUE, pattern = "*.csv")
)

# Initialize empty dataframes
alldata <- data.frame()
alldata_mist <- data.frame()
alldata_mocri <- data.frame()

# Progress bar
progbar <- progress_bar$new(total = length(files))

# Loop through files
for (file in files) {
  progbar$tick()

  # Read the CSV file
  data <- read.csv(file)

  # Browser info
  browser <- data[data$screen == "browser_info", ][1, ]

  # Skip if no valid Prolific ID
  if(is.na(browser$researcher)) {
    if(is.na(browser$prolific_id)) {
      next
    } else {
      browser$researcher <- "Prolific"
    }
  } else if(browser$researcher == "test") {
    next
  }

  # Participant dataframe
  df <- data.frame(
    Participant = gsub(".csv", "", basename(file)),
    # Prolific_ID = browser$prolific_id,
    Sample = browser$researcher,
    Experiment_Duration = max(data$time_elapsed) / 1000 / 60,
    Date = browser$date,
    Time = browser$time,
    Browser = browser$browser,
    Mobile = browser$mobile,
    Platform = browser$os,
    Screen_Width = browser$screen_width,
    Screen_Height = browser$screen_height
  )

  # Demographics 1
  demo1 <- jsonlite::fromJSON(data[data$screen == "demographics_1", "response"])
  df$Gender <- demo1$gender

  # Education cleaning
  edu <- demo1$education
  df$Education <- case_when(
    str_detect(tolower(edu), "bachelor") ~ "Bachelor",
    str_detect(tolower(edu), "master") ~ "Master",
    str_detect(tolower(edu), "doctorate") ~ "Doctorate",
    str_detect(tolower(edu), "high school") ~ "High School",
    .default = edu
  )


  # Demographics 2
  demo2 <- jsonlite::fromJSON(data[data$screen == "demographics_2", "response"])
  if(str_detect(demo2$age, "66d633|66f418c")) {
    df$Age <- NA
  } else {
    df$Age <- as.numeric(demo2$age)
  }
  if (!is.na(df$Age) && df$Age == 1964) df$Age <- 60
  df$Ethnicity <- demo2$ethnicity

  df$COVID_Vaccination <- demo1$vaccination_status
  df$FakeNews_SubjectiveDiscrimination <- demo1$belief_likelihood_status


  # BFI (Big Five Inventory)
  bfi <- data[data$screen == "questionnaire_BFI10", ]
  df$BFI_Duration <- bfi$rt / 1000 / 60
  bfi_resp <- jsonlite::fromJSON(bfi$response)
  for (item in names(bfi_resp)) {
    df[[item]] <- as.numeric(bfi_resp[[item]])
  }

  # VSA (Vulnerability to Stress Assessment)
  vsa <- data[data$screen == "questionnaire_VSA", ]
  df$VSA_Duration <- vsa$rt / 1000 / 60
  vsa_resp <- jsonlite::fromJSON(vsa$response)
  for (item in names(vsa_resp)) {
    df[[item]] <- as.numeric(vsa_resp[[item]])
  }

  # ANES (American National Election Studies)
  anes <- data[data$screen == "questionnaire_ANES", ]
  df$ANES_Duration <- anes$rt / 1000 / 60
  anes_resp <- jsonlite::fromJSON(anes$response)
  df$Political_Ideology <- str_remove(anes_resp$ANES_1, ".*\\. ")
  df$Political_Ideology <- ifelse(str_detect(df$Political_Ideology, "Moderate"), "Moderate", df$Political_Ideology)
  df$Political_Ideology <- ifelse(str_detect(df$Political_Ideology, "slightly C"), "Slightly Conservative", df$Political_Ideology)
  df$Political_Ideology <- ifelse(str_detect(df$Political_Ideology, "Don't Know"), "Don't Know", df$Political_Ideology)
  df$Political_Affiliation <- str_remove(anes_resp$ANES_2, ".*\\. ")
  df$Political_Affiliation <- ifelse(str_detect(df$Political_Affiliation, "Democratic"), "Democrat", df$Political_Affiliation)
  df$Political_Affiliation <- ifelse(str_detect(df$Political_Affiliation, "Republican"), "Republican", df$Political_Affiliation)
  df$Political_Affiliation <- ifelse(str_detect(df$Political_Affiliation, "'independent'"), "None or Independent", df$Political_Affiliation)
  df$Political_Affiliation <- ifelse(df$Political_Affiliation == "Other" & df$Sample == "Prolific", "None or Independent", df$Political_Affiliation)

  # GCBS15 (Generic Conspiracist Beliefs Scale)
  gcbs <- data[data$screen == "questionnaire_GCBS15", ][1, ]
  df$GCBS_Duration <- gcbs$rt / 1000 / 60
  gcbs_resp <- jsonlite::fromJSON(gcbs$response)
  for (item in names(gcbs_resp)) {
    df[[item]] <- as.numeric(gcbs_resp[[item]])
  }

  # Veracity confidence
  confidence <- data[data$screen == "veracity_confidence", ]
  df$Intervention_Questions_Confidence <- jsonlite::fromJSON(confidence$response)$confidence

  # Badnews/Tetris questions
  intervention <- data[data$screen == "intervention", ]
  df$Intervention <- ifelse(str_detect(intervention$stimulus, "Tetris"), "Tetris", "BadNewsGame")
  df$Intervention_Duration <- intervention$rt / 1000 / 60
  if (nrow(data[data$screen == "questionnaire_badnews", ]) == 0) {
    taskQ <- data[data$screen == "questionnaire_tetris", ][1, ]
    taskQ_resp <- jsonlite::fromJSON(taskQ$response)
    df$Intervention_Questions_Score <- taskQ_resp$high_score
    if(df$Intervention != "Tetris") stop()
  } else {
    taskQ <- data[data$screen == "questionnaire_badnews", ][1, ]
    taskQ_resp <- jsonlite::fromJSON(taskQ$response)
    df$Intervention_Questions_Score <- taskQ_resp$followers
  }
  df$Intervention_Questions_Duration <- taskQ$rt / 1000 / 60
  df$Intervention_Questions_Favorite <- taskQ_resp$favorite_part
  df$Intervention_Questions_Repeat <- taskQ_resp$play_again
  df$Intervention_Questions_Recommend <- taskQ_resp$recommend


  # MOCRI BOLD
  mocri_bold <- data[data$screen == "questionnaire_MOCRI_BOLD_12", ]
  mocri_bold_resp <- jsonlite::fromJSON(mocri_bold$response)
  for (item in names(mocri_bold_resp)) {
    df[[item]] <- as.numeric(mocri_bold_resp[[item]])
  }
  mocri_nonbold <- data[data$screen == "questionnaire_MOCRI_NONBOLD_12", ]
  # df$MOCRI_NONBOLD_Duration <- mocri_nonbold$rt / 1000 / 60
  mocri_nonbold_resp <- jsonlite::fromJSON(mocri_nonbold$response)
  for (item in names(mocri_nonbold_resp)) {
    df[[item]] <- as.numeric(mocri_nonbold_resp[[item]])
  }
  names(df) <- gsub("MOCRI_", "MOCRI_POST_", names(df))
  names(df) <- gsub("MOCRI_POST_BOLD", "MOCRI_PRE", names(df))


  # MIST (Pre and Post)
  # Pre
  mist_pre <- data[data$screen == "questionnaire_mist_pre", ]
  # df$MIST_Duration_Pre <- mist_pre$rt / 1000 / 60
  mist_pre <- jsonlite::fromJSON(mist_pre$response)
  df_mist_pre <- data.frame(
    Participant = df$Participant,
    Item = names(mist_pre),
    Realness = as.numeric(unlist(mist_pre)) / 100,
    Condition = "Pretest"
  )

  # Post
  mist_post <- data[data$screen == "questionnaire_mist_post", ]
  # df$MIST_Duration_Post <- mist_post$rt / 1000 / 60
  mist_post <- jsonlite::fromJSON(mist_post$response)
  df_mist_post <- data.frame(
    Participant = df$Participant,
    Item = names(mist_post),
    Realness = as.numeric(unlist(mist_post)) / 100,
    Condition = "Posttest"
  )

  # Combine MIST pre and post
  df_mist <- rbind(df_mist_pre, df_mist_post)
  df_mist$Nature <- tools::toTitleCase(stringr::str_split(df_mist$Item, "_", simplify = TRUE)[, 2])
  df_mist$Topic <- tools::toTitleCase(stringr::str_split(df_mist$Item, "_", simplify = TRUE)[, 3])

  # Add to alldata
  alldata <- rbind(alldata, df)
  alldata_mist <- rbind(alldata_mist, df_mist)
}

# How many
table(alldata$Sample)
alldata$Sample <- ifelse(alldata$Sample %in% c("Prolific", "snow"), "USA", "UK")


# Anonymize
alldata$Prolific_ID <- NULL
ids <- paste0("S", sprintf("%03d", 1:nrow(alldata)))
names(ids) <- alldata$Participant
alldata$Participant <- ids[alldata$Participant]
alldata_mist$Participant <- ids[alldata_mist$Participant]

# Save results
write.csv(alldata, "../data/rawdata_participants.csv", row.names = FALSE)
write.csv(alldata_mist, "../data/rawdata_mist.csv", row.names = FALSE)
